src/stdpar/plugin-SiPixelClusterizer/gpuClustering.h

```diff
@@ -1,292 +1,262 @@
-        atomicMin(&msize, i);
-        break;
+      // skip threads not associated to an existing pixel
+      for (int i = first; i < numElements; i++) {
+        if (id[i] == InvId)  // skip invalid pixels
+          continue;
+        if (id[i] != thisModuleId) {  // find the first pixel in a different module
+          cms::cuda::atomicMin(&msize, i);
+          break;
+        }
       }
-    }
 
-    //init hist  (ymax=416 < 512 : 9bits)
-    constexpr uint32_t maxPixInModule = 4000;
-    constexpr auto nbins = phase1PixelTopology::numColsInModule + 2;  //2+2;
-    using Hist = cms::cuda::HistoContainer<uint16_t, nbins, maxPixInModule, 9, uint16_t>;
-    __shared__ Hist hist;
-    __shared__ typename Hist::Counter ws[32];
-    for (auto j = threadIdx.x; j < Hist::totbins(); j += blockDim.x) {
-      hist.off[j] = 0;
-    }
-    __syncthreads();
+      //init hist  (ymax=416 < 512 : 9bits)
+      constexpr uint32_t maxPixInModule = 4000;
+      constexpr auto nbins = phase1PixelTopology::numColsInModule + 2;  //2+2;
+      using Hist = cms::cuda::HistoContainer<uint16_t, nbins, maxPixInModule, 9, uint16_t>;
+      Hist hist;
+      for (auto j = 0; j < Hist::totbins(); ++j) {
+        hist.off[j] = 0;
+      }
 
-    assert((msize == numElements) or ((msize < numElements) and (id[msize] != thisModuleId)));
+      assert((msize == numElements) or ((msize < numElements) and (id[msize] != thisModuleId)));
 
-    // limit to maxPixInModule  (FIXME if recurrent (and not limited to simulation with low threshold) one will need to implement something cleverer)
-    if (0 == threadIdx.x) {
+      // limit to maxPixInModule  (FIXME if recurrent (and not limited to simulation with low threshold) one will need to implement something cleverer)
       if (msize - firstPixel > maxPixInModule) {
         printf("too many pixels in module %d: %d > %d\n", thisModuleId, msize - firstPixel, maxPixInModule);
         msize = maxPixInModule + firstPixel;
       }
-    }
 
-    __syncthreads();
-    assert(msize - firstPixel <= maxPixInModule);
+      assert(msize - firstPixel <= maxPixInModule);
 
 #ifdef GPU_DEBUG
-    __shared__ uint32_t totGood;
-    totGood = 0;
-    __syncthreads();
+      uint32_t totGood;
+      totGood = 0;
 #endif
 
     // fill histo
-    for (int i = first; i < msize; i += blockDim.x) {
-      if (id[i] == InvId)  // skip invalid pixels
-        continue;
-      hist.count(y[i]);
+      for (int i = first; i < msize; ++i) {
+        if (id[i] == InvId)  // skip invalid pixels
+          continue;
+        hist.count(y[i]);
 #ifdef GPU_DEBUG
-      atomicAdd(&totGood, 1);
+        std::atomic_ref<uint32_t> inc{totGood};
+        ++inc;
 #endif
-    }
-    __syncthreads();
-    if (threadIdx.x < 32)
-      ws[threadIdx.x] = 0;  // used by prefix scan...
-    __syncthreads();
-    hist.finalize(ws);
-    __syncthreads();
+      }
+      hist.finalizeSeq();
```

Comment:
hist.finalize() calls a parallel algorithm, we need a special version to be called withing a for_each lambda